### PR TITLE
Implement latency stats command in JAVA API

### PR DESCRIPTION
**Pull request by ericsson-peiwuyang**
Implement latency stats command in JAVA API.

### CODE REVIEW

File: src/main/java/com/cisco/trex/stateful/model/stats/LatencyStats.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+     *       "m_unsup_prot": 0,
+     *       "tx_arp": 0
+     *     }
+     *   } ...
+     * }
+     * </pre>
+     * see http://trex-tgn.cisco.com/trex/doc/cp_docs/api/json_fields.html#trex-latecny-field for detail description.
+     */
+    @JsonProperty("data")
+    public Map data;
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/LatencyStats.java]: this should be defined to a java bean and all its internal properties as well, like:

```java
public LatencyData data;
public String name;
public Integer type;
```
then create LatencyData.java
```java
public double cpuUtil;
......
public Map<Integer, PortLatencyData>;
```

then implement the next hierarchy java bean for "PortLatencyData" and recursively.

Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/LatencyStats.java]: Ok, I have adjusted the code, thanks.

File: src/main/java/com/cisco/trex/stateful/model/stats/HistogramData.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    @JsonProperty("val")
+    public Integer val;
+
+    public Integer getKey() {
+        return key;
+    }
+
+    public void setKey(Integer key) {
+        this.key = key;
+    }
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/HistogramData.java]: it's better to add @JsonProperty(key) annotation for all set/getXXX if it has corresponding field in json response.

File: src/main/java/com/cisco/trex/stateful/model/stats/LatencyData.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    public String cpu_util;
+
+    @JsonProperty("epoch")
+    public Integer epoch;
+
+    public Map<Integer, PortLatencyData> portLatencyDataMap;
+
+    @JsonProperty("unknown")
+    public Integer unknown;
+
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/LatencyData.java]: it's better to add @JsonProperty(key) annotation for all set/getXXX if it has corresponding field in json response.

File: src/main/java/com/cisco/trex/stateful/model/stats/LatencyData.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Map;
+
+public class LatencyData {
+
+    @JsonProperty("cpu_util")
+    public String cpu_util;
+
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/LatencyData.java]: naming with java convention: cpuUtil

File: src/main/java/com/cisco/trex/stateful/model/stats/LatencyStats.java [hunk 5]
```diff
@@ -45,19 +45,19 @@
...
+    public LatencyData data;
 
     @JsonProperty("name")
     public String name;
 
     @JsonProperty("type")
     public Integer type;
 
-    public Map getData() {
+    public LatencyData getData() {
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/LatencyStats.java]: it's better to add @JsonProperty(key) annotation for all set/getXXX if it has corresponding field in json response.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+import java.util.List;
+
+public class PortHistData {
+
+    @JsonProperty("cnt")
+    public Integer cnt;
+
+    @JsonProperty("high_cnt")
+    public Integer high_cnt;
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java]: naming with java convention highCnt.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+    @JsonProperty("histogram")
+    public List<HistogramData> histogram;
+
+    @JsonProperty("max_usec")
+    public Integer max_usec;
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java]: naming with java convention

File: src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+    @JsonProperty("min_usec")
+    public Integer min_usec;
+
+    @JsonProperty("s_avg")
+    public Double s_avg;
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java]: naming with java convention

File: src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+    @JsonProperty("s_max")
+    public Double s_max;
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java]: naming with java convention

File: src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
+package com.cisco.trex.stateful.model.stats;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.List;
+
+public class PortHistData {
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java]: rename to "LatencyPortHist"

File: src/main/java/com/cisco/trex/stateful/model/stats/HistogramData.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
+package com.cisco.trex.stateful.model.stats;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class HistogramData {
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/HistogramData.java]: can make it inner class of LatencyPortHist and rename it with "histogram", then all names can keep alignment at most with original json format

File: src/main/java/com/cisco/trex/stateful/model/stats/PortLatencyData.java [hunk 12]
```diff
@@ -0,0 +1,28 @@
+package com.cisco.trex.stateful.model.stats;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class PortLatencyData {
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/PortLatencyData.java]: rename to "LatencyPortData"

File: src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    @JsonProperty("min_usec")
+    public Integer min_usec;
+
+    @JsonProperty("s_avg")
+    public Double s_avg;
+
+    @JsonProperty("s_max")
+    public Double s_max;
+
+    public Integer getCnt() {
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java]: it's better to add @JsonProperty(key) annotation for all set/getXXX if it has corresponding field in json response.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortLatencyData.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+    @JsonProperty("hist")
+    public PortHistData hist;
+
+    @JsonProperty("stats")
+    public PortStatsData stats;
+
+    public PortHistData getHist() {
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/PortLatencyData.java]: it's better to add @JsonProperty(key) annotation for all set/getXXX if it has corresponding field in json response.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortStatsData.java [hunk 15]
```diff
@@ -0,0 +1,182 @@
+package com.cisco.trex.stateful.model.stats;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class PortStatsData {
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/PortStatsData.java]: rename to "LatencyPortStats"

File: src/main/java/com/cisco/trex/stateful/model/stats/PortStatsData.java [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+
+    @JsonProperty("ignore_bytes")
+    public Integer ignore_bytes;
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/PortStatsData.java]: naming with java convention, same comments for all below fields

File: src/main/java/com/cisco/trex/stateful/model/stats/PortStatsData.java [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+    @JsonProperty("m_tx_pkt_ok")
+    public Integer m_tx_pkt_ok;
+
+    @JsonProperty("m_unsup_prot")
+    public Integer m_unsup_prot;
+
+    @JsonProperty("tx_arp")
+    public Integer tx_arp;
+
+    public Integer getIgnore_bytes() {
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/model/stats/PortStatsData.java]: it's better to add @JsonProperty(key) annotation for all set/getXXX if it has corresponding field in json response.

File: src/main/java/com/cisco/trex/stateful/TRexAstfClient.java [hunk 18]
```diff
@@ -350,7 +353,25 @@ private MetaData getAstfStatsMetaData() {
      */
     public LatencyStats getLatencyStats() {
         Map<String, Object> payload = this.createPayload();
-        return this.callMethod("get_latency_stats", payload, LatencyStats.class).get();
+        String json = this.callMethod("get_latency_stats", payload);
+        JsonElement response = new JsonParser().parse(json);
+        JsonElement latencyStatsData = response.getAsJsonArray().get(0).getAsJsonObject().get("result");
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/TRexAstfClient.java]: confusing name with below java classes definition, can this be "latencyStatsJsonElement"?

File: src/main/java/com/cisco/trex/stateful/TRexAstfClient.java [hunk 19]
```diff
@@ -350,7 +353,25 @@ private MetaData getAstfStatsMetaData() {
...
+        //only can parse a part of data, PortLatencyData need to be parsed manually.
+        LatencyStats latencyStats = GSON.fromJson(latencyStatsData, LatencyStats.class);
+        JsonElement dataElement = latencyStatsData.getAsJsonObject().get("data");
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/TRexAstfClient.java]: can name with latencyDataJsonElement

File: src/main/java/com/cisco/trex/stateful/TRexAstfClient.java [hunk 20]
```diff
@@ -350,7 +353,25 @@ private MetaData getAstfStatsMetaData() {
...
+        JsonObject data = dataElement.getAsJsonObject();
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/TRexAstfClient.java]: name with "latencyDataJsonObject"

File: src/main/java/com/cisco/trex/stateful/TRexAstfClient.java [hunk 21]
```diff
@@ -350,7 +353,25 @@ private MetaData getAstfStatsMetaData() {
...
+        Map<Integer, PortLatencyData> portLatencyDataMap = new HashMap<>();
+        // parse PortLatencyData manually
+        for (Map.Entry<String, JsonElement> entry : data.entrySet()) {
+            String jsonKey = entry.getKey();
+            if(jsonKey.startsWith("port")){
+                Integer port = Integer.parseInt(jsonKey.substring(5));
```
Comment (leomaatEric) [inline src/main/java/com/cisco/trex/stateful/TRexAstfClient.java]: portIndex

File: src/main/java/com/cisco/trex/stateful/model/stats/HistogramData.java (hunk 2 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/HistogramData.java]: ok, I will adjust it.

File: src/main/java/com/cisco/trex/stateful/model/stats/LatencyData.java (hunk 3 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/LatencyData.java]: ok, I will adjust it.

File: src/main/java/com/cisco/trex/stateful/model/stats/LatencyData.java (hunk 4 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/LatencyData.java]: ok, I will change it to camel case.

File: src/main/java/com/cisco/trex/stateful/model/stats/LatencyStats.java (hunk 5 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/LatencyStats.java]: ok, I will adjust it.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java (hunk 6 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java]: ok, I will change it to camel case.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java (hunk 7 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java]: ok, I will change it to camel case.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java (hunk 8 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java]: ok, I will change it to camel case.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java (hunk 9 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java]: ok, I will change it to camel case.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java (hunk 10 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java]: ok, I will rename it.

File: src/main/java/com/cisco/trex/stateful/model/stats/HistogramData.java (hunk 11 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/HistogramData.java]: good suggestion, thanks.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortLatencyData.java (hunk 12 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/PortLatencyData.java]: ok, I will rename it.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java (hunk 13 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/PortHistData.java]: ok

File: src/main/java/com/cisco/trex/stateful/model/stats/PortLatencyData.java (hunk 14 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/PortLatencyData.java]: ok, I will add it.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortStatsData.java (hunk 15 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/PortStatsData.java]: ok, I will rename it.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortStatsData.java (hunk 16 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/PortStatsData.java]: ok, I will change it to camel case.

File: src/main/java/com/cisco/trex/stateful/model/stats/PortStatsData.java (hunk 17 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/model/stats/PortStatsData.java]: ok, I will add it.

File: src/main/java/com/cisco/trex/stateful/TRexAstfClient.java (hunk 18 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/TRexAstfClient.java]: yes, I will rename this variable.

File: src/main/java/com/cisco/trex/stateful/TRexAstfClient.java (hunk 19 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/TRexAstfClient.java]: yes, I will rename this variable.

File: src/main/java/com/cisco/trex/stateful/TRexAstfClient.java (hunk 20 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/TRexAstfClient.java]: yes, I will rename this variable.

File: src/main/java/com/cisco/trex/stateful/TRexAstfClient.java (hunk 21 above)
Comment (ericsson-peiwuyang) [reply src/main/java/com/cisco/trex/stateful/TRexAstfClient.java]: yes, I will rename this variable.

**Pull request merged**